File: android/src/main/kotlin/com/zcreations/executorch_flutter/ExecutorchLifecycleManager.kt
/**
 * ExecuTorch Lifecycle Manager - Android Platform Resource Management
 *
 * This class manages the lifecycle of ExecuTorch resources on Android, including
 * application state monitoring, memory pressure handling, and proper cleanup
 * during app transitions. It ensures optimal resource usage and prevents
 * memory leaks in the Android environment.
 *
 * Features:
 * - Application lifecycle monitoring (background, foreground, low memory)
 * - Memory pressure detection and response
 * - Automatic model disposal during memory warnings
 * - Resource cleanup coordination
 * - Performance monitoring and reporting
 */
package com.zcreations.executorch_flutter

import android.app.Activity
import android.app.Application
import android.content.ComponentCallbacks2
import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.util.Log
import kotlinx.coroutines.*
import java.lang.ref.WeakReference
import java.util.concurrent.ConcurrentHashMap

// Import Pigeon generated API
import com.zcreations.executorch_flutter.generated.*

/**
 * Manages ExecuTorch resource lifecycle in response to Android app state changes
 */
class ExecutorchLifecycleManager private constructor(
    private val application: Application
) : Application.ActivityLifecycleCallbacks, ComponentCallbacks2 {

    companion object {
        private const val TAG = "ExecutorchLifecycleManager"
        private const val MEMORY_WARNING_COOLDOWN_MS = 10_000L // 10 seconds

        @Volatile
        private var INSTANCE: ExecutorchLifecycleManager? = null

        fun getInstance(application: Application): ExecutorchLifecycleManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: ExecutorchLifecycleManager(application).also {
                    INSTANCE = it
                    application.registerActivityLifecycleCallbacks(it)
                    application.registerComponentCallbacks(it)
                }
            }
        }
    }

    // Lifecycle state
    private var isInForeground = false
    private var activeActivityCount = 0
    private var lastMemoryWarning = 0L

    // Model managers to coordinate with
    private val modelManagers = ConcurrentHashMap<String, WeakReference<ExecutorchModelManager>>()

    // Coroutine scope for lifecycle operations
    private val lifecycleScope = CoroutineScope(
        Dispatchers.Default + SupervisorJob() + CoroutineName("ExecutorchLifecycle")
    )

    init {
        Log.d(TAG, "ExecutorchLifecycleManager initialized")
    }

    // MARK: - Public API

    /**
     * Register a model manager for lifecycle coordination
     */
    fun registerModelManager(id: String, manager: ExecutorchModelManager) {
        // Clean up any deallocated references
        cleanupDeadReferences()

        // Add new manager
        modelManagers[id] = WeakReference(manager)
        Log.d(TAG, "Registered model manager: $id, total: ${modelManagers.size}")
    }

    /**
     * Unregister a model manager
     */
    fun unregisterModelManager(id: String) {
        modelManagers.remove(id)
        Log.d(TAG, "Unregistered model manager: $id, remaining: ${modelManagers.size}")
    }

    /**
     * Force cleanup of all resources
     */
    suspend fun forceCleanup() {
        Log.d(TAG, "Force cleanup requested")
        disposeAllModels()
    }

    /**
     * Get current memory usage statistics
     */
    fun getMemoryStats(): AndroidMemoryStats {
        val runtime = Runtime.getRuntime()
        val activityManager = application.getSystemService(Context.ACTIVITY_SERVICE) as android.app.ActivityManager
        val memInfo = android.app.ActivityManager.MemoryInfo()
        activityManager.getMemoryInfo(memInfo)

        return AndroidMemoryStats(
            usedBytes = runtime.totalMemory() - runtime.freeMemory(),
            maxBytes = runtime.maxMemory(),
            availableBytes = memInfo.availMem,
            totalSystemBytes = memInfo.totalMem,
            isLowMemory = memInfo.lowMemory,
            lastWarningTime = lastMemoryWarning
        )
    }

    /**
     * Check if app is currently in foreground
     */
    fun isAppInForeground(): Boolean = isInForeground

    // MARK: - Activity Lifecycle Callbacks

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
        Log.d(TAG, "Activity created: ${activity.localClassName}")
    }

    override fun onActivityStarted(activity: Activity) {
        activeActivityCount++
        updateForegroundState()
        Log.d(TAG, "Activity started: ${activity.localClassName}, active count: $activeActivityCount")
    }

    override fun onActivityResumed(activity: Activity) {
        Log.d(TAG, "Activity resumed: ${activity.localClassName}")
    }

    override fun onActivityPaused(activity: Activity) {
        Log.d(TAG, "Activity paused: ${activity.localClassName}")
    }

    override fun onActivityStopped(activity: Activity) {
        activeActivityCount--
        updateForegroundState()
        Log.d(TAG, "Activity stopped: ${activity.localClassName}, active count: $activeActivityCount")

        if (activeActivityCount == 0) {
            handleAppEnteredBackground()
        }
    }

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {
        Log.d(TAG, "Activity saving state: ${activity.localClassName}")
    }

    override fun onActivityDestroyed(activity: Activity) {
        Log.d(TAG, "Activity destroyed: ${activity.localClassName}")
    }

    // MARK: - Component Callbacks

    override fun onConfigurationChanged(newConfig: Configuration) {
        Log.d(TAG, "Configuration changed")
    }

    override fun onLowMemory() {
        Log.w(TAG, "System low memory warning")
        handleMemoryPressure(ComponentCallbacks2.TRIM_MEMORY_COMPLETE)
    }

    override fun onTrimMemory(level: Int) {
        Log.w(TAG, "Memory trim requested, level: $level")
        handleMemoryPressure(level)
    }

    // MARK: - Private Implementation

    private fun updateForegroundState() {
        val wasInForeground = isInForeground
        isInForeground = activeActivityCount > 0

        if (wasInForeground != isInForeground) {
            if (isInForeground) {
                handleAppEnteredForeground()
            }
        }
    }

    private fun handleAppEnteredForeground() {
        Log.d(TAG, "App entered foreground")
        // Restore any paused operations if needed
    }

    private fun handleAppEnteredBackground() {
        Log.d(TAG, "App entered background")

        // TODO: Consider less aggressive background cleanup for demo apps
        // For now, skip automatic model disposal to prevent user frustration
        // lifecycleScope.launch {
        //     performBackgroundCleanup()
        // }
    }

    private fun handleMemoryPressure(level: Int) {
        val now = System.currentTimeMillis()

        // Avoid too frequent memory warning handling
        if (now - lastMemoryWarning < MEMORY_WARNING_COOLDOWN_MS) {
            Log.d(TAG, "Memory pressure ignored (too frequent)")
            return
        }

        lastMemoryWarning = now
        Log.w(TAG, "Memory pressure level: $level")

        lifecycleScope.launch {
            when {
                level >= ComponentCallbacks2.TRIM_MEMORY_COMPLETE -> {
                    // Critical memory pressure - dispose most models
                    handleCriticalMemoryPressure()
                }
                level >= ComponentCallbacks2.TRIM_MEMORY_MODERATE -> {
                    // Moderate memory pressure - dispose some models
                    handleModerateMemoryPressure()
                }
                level >= ComponentCallbacks2.TRIM_MEMORY_BACKGROUND -> {
                    // App in background - conservative cleanup
                    handleBackgroundMemoryPressure()
                }
                else -> {
                    // Light memory pressure - minimal cleanup
                    handleLightMemoryPressure()
                }
            }
        }
    }

    private suspend fun handleCriticalMemoryPressure() {
        Log.w(TAG, "Handling critical memory pressure")

        val activeManagers = getActiveManagers()
        for (manager in activeManagers) {
            manager.handleMemoryPressure(severity = MemoryPressureSeverity.CRITICAL)
        }

        // Force garbage collection
        System.gc()
    }

    private suspend fun handleModerateMemoryPressure() {
        Log.w(TAG, "Handling moderate memory pressure")

        val activeManagers = getActiveManagers()
        for (manager in activeManagers) {
            manager.handleMemoryPressure(severity = MemoryPressureSeverity.MODERATE)
        }
    }

    private suspend fun handleBackgroundMemoryPressure() {
        Log.d(TAG, "Handling background memory pressure")

        val activeManagers = getActiveManagers()
        for (manager in activeManagers) {
            manager.handleMemoryPressure(severity = MemoryPressureSeverity.BACKGROUND)
        }
    }

    private suspend fun handleLightMemoryPressure() {
        Log.d(TAG, "Handling light memory pressure")

        val activeManagers = getActiveManagers()
        for (manager in activeManagers) {
            manager.handleMemoryPressure(severity = MemoryPressureSeverity.LIGHT)
        }
    }

    private suspend fun performBackgroundCleanup() {
        Log.d(TAG, "Performing background cleanup")

        val activeManagers = getActiveManagers()
        for (manager in activeManagers) {
            manager.performBackgroundCleanup()
        }
    }

    private suspend fun disposeAllModels() {
        Log.d(TAG, "Disposing all models")

        val activeManagers = getActiveManagers()
        for (manager in activeManagers) {
            manager.disposeAllModels()
        }

        modelManagers.clear()
    }

    private fun getActiveManagers(): List<ExecutorchModelManager> {
        cleanupDeadReferences()
        return modelManagers.values.mapNotNull { it.get() }
    }

    private fun cleanupDeadReferences() {
        val deadKeys = modelManagers.entries
            .filter { it.value.get() == null }
            .map { it.key }

        deadKeys.forEach { modelManagers.remove(it) }

        if (deadKeys.isNotEmpty()) {
            Log.d(TAG, "Cleaned up ${deadKeys.size} dead model manager references")
        }
    }

    fun cleanup() {
        lifecycleScope.cancel()
        application.unregisterActivityLifecycleCallbacks(this)
        application.unregisterComponentCallbacks(this)
        Log.d(TAG, "ExecutorchLifecycleManager cleaned up")
    }
}

// MARK: - Extension for Model Manager Integration

/**
 * Memory pressure severity levels
 */
enum class MemoryPressureSeverity {
    LIGHT,      // Minimal cleanup needed
    BACKGROUND, // App in background, moderate cleanup
    MODERATE,   // Dispose some models
    CRITICAL    // Dispose most models immediately
}

/**
 * Extension functions for ExecutorchModelManager to handle lifecycle events
 */
suspend fun ExecutorchModelManager.handleMemoryPressure(severity: MemoryPressureSeverity) {
    Log.d("ExecutorchModelManager", "Handling memory pressure: $severity")

    when (severity) {
        MemoryPressureSeverity.CRITICAL -> {
            // Dispose 75% of models
            val loadedIds = getLoadedModelIds()
            val modelsToDispose = loadedIds.take(loadedIds.size * 3 / 4)
            disposeModels(modelsToDispose)
        }
        MemoryPressureSeverity.MODERATE -> {
            // Dispose 50% of models
            val loadedIds = getLoadedModelIds()
            val modelsToDispose = loadedIds.take(loadedIds.size / 2)
            disposeModels(modelsToDispose)
        }
        MemoryPressureSeverity.BACKGROUND -> {
            // Dispose 25% of models
            val loadedIds = getLoadedModelIds()
            val modelsToDispose = loadedIds.take(loadedIds.size / 4)
            disposeModels(modelsToDispose)
        }
        MemoryPressureSeverity.LIGHT -> {
            // Just request GC
            TensorMemoryManager.requestGCIfNeeded(threshold = 0.7)
        }
    }
}

suspend fun ExecutorchModelManager.performBackgroundCleanup() {
    Log.d("ExecutorchModelManager", "Performing background cleanup")
    handleMemoryPressure(MemoryPressureSeverity.BACKGROUND)
}

private suspend fun ExecutorchModelManager.disposeModels(modelIds: List<String>) {
    for (modelId in modelIds) {
        try {
            disposeModel(modelId)
            Log.d("ExecutorchModelManager", "Disposed model due to memory pressure: $modelId")
        } catch (e: Exception) {
            Log.e("ExecutorchModelManager", "Failed to dispose model during memory pressure: $modelId", e)
        }
    }
}

// MARK: - Supporting Data Types

/**
 * Android-specific memory usage statistics
 */
data class AndroidMemoryStats(
    val usedBytes: Long,
    val maxBytes: Long,
    val availableBytes: Long,
    val totalSystemBytes: Long,
    val isLowMemory: Boolean,
    val lastWarningTime: Long
) {
    val usageMB: Double
        get() = usedBytes.toDouble() / (1024 * 1024)

    val maxMB: Double
        get() = maxBytes.toDouble() / (1024 * 1024)

    val availableMB: Double
        get() = availableBytes.toDouble() / (1024 * 1024)

    val usagePercentage: Double
        get() = if (maxBytes > 0) usedBytes.toDouble() / maxBytes.toDouble() * 100 else 0.0

    val systemUsagePercentage: Double
        get() = if (totalSystemBytes > 0) (totalSystemBytes - availableBytes).toDouble() / totalSystemBytes.toDouble() * 100 else 0.0
}

File: macos/executorch_flutter/Sources/executorch_flutter/ExecutorchLifecycleManager.swift
/**
 * ExecuTorch Lifecycle Manager - iOS Platform Resource Management
 *
 * This class manages the lifecycle of ExecuTorch resources on iOS, including
 * application state monitoring, memory pressure handling, and proper cleanup
 * during app transitions. It ensures optimal resource usage and prevents
 * memory leaks in the iOS environment.
 *
 * Features:
 * - Application lifecycle monitoring (background, foreground, termination)
 * - Memory pressure detection and response
 * - Automatic model disposal during memory warnings
 * - Resource cleanup coordination
 * - Performance monitoring and reporting
 */
import Foundation
#if os(iOS)
import UIKit
#elseif os(macOS)
import AppKit
#endif

/**
 * Manages ExecuTorch resource lifecycle in response to iOS app state changes
 */
class ExecutorchLifecycleManager {

    fileprivate static let TAG = "ExecutorchLifecycleManager"

    // Singleton instance
    static let shared = ExecutorchLifecycleManager()

    // Lifecycle state
    private var isActive = false
    #if os(iOS)
    private var backgroundTaskIdentifier: UIBackgroundTaskIdentifier = .invalid
    #endif

    // Model managers to coordinate with
    private var modelManagers: [WeakModelManagerRef] = []

    // Memory monitoring
    private var lastMemoryWarning: Date?
    private let memoryWarningCooldown: TimeInterval = 10.0 // seconds

    // Notification observers
    private var notificationObservers: [NSObjectProtocol] = []

    /**
     * Weak reference wrapper for model managers
     */
    private struct WeakModelManagerRef {
        weak var manager: ExecutorchModelManager?
    }

    private init() {
        setupApplicationLifecycleObservers()
        setupMemoryPressureObservers()
        print("[\(Self.TAG)] ExecutorchLifecycleManager initialized")
    }

    deinit {
        cleanup()
        print("[\(Self.TAG)] ExecutorchLifecycleManager deinitialized")
    }

    // MARK: - Public API

    /**
     * Register a model manager for lifecycle coordination
     */
    func registerModelManager(_ manager: ExecutorchModelManager) {
        // Clean up any deallocated references
        modelManagers.removeAll { $0.manager == nil }

        // Add new manager
        modelManagers.append(WeakModelManagerRef(manager: manager))
        print("[\(Self.TAG)] Registered model manager, total: \(modelManagers.count)")
    }

    /**
     * Unregister a model manager
     */
    func unregisterModelManager(_ manager: ExecutorchModelManager) {
        modelManagers.removeAll { $0.manager === manager }
        print("[\(Self.TAG)] Unregistered model manager, remaining: \(modelManagers.count)")
    }

    /**
     * Force cleanup of all resources
     */
    func forceCleanup() async {
        print("[\(Self.TAG)] Force cleanup requested")
        await disposeAllModels()
    }

    /**
     * Get current memory usage statistics
     */
    func getMemoryStats() -> MemoryStats {
        let usedMemory = getCurrentMemoryUsage()
        let availableMemory = ProcessInfo.processInfo.physicalMemory
        let memoryPressure = getMemoryPressureLevel()

        return MemoryStats(
            usedBytes: usedMemory,
            availableBytes: availableMemory,
            pressureLevel: memoryPressure,
            lastWarningTime: lastMemoryWarning
        )
    }

    // MARK: - Application Lifecycle Handling

    private func setupApplicationLifecycleObservers() {
        #if os(iOS)
        let notificationCenter = NotificationCenter.default

        // App will become active
        let willBecomeActiveObserver = notificationCenter.addObserver(
            forName: UIApplication.willEnterForegroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppWillBecomeActive()
        }
        notificationObservers.append(willBecomeActiveObserver)

        // App did become active
        let didBecomeActiveObserver = notificationCenter.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppDidBecomeActive()
        }
        notificationObservers.append(didBecomeActiveObserver)

        // App will resign active
        let willResignActiveObserver = notificationCenter.addObserver(
            forName: UIApplication.willResignActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppWillResignActive()
        }
        notificationObservers.append(willResignActiveObserver)

        // App did enter background
        let didEnterBackgroundObserver = notificationCenter.addObserver(
            forName: UIApplication.didEnterBackgroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppDidEnterBackground()
        }
        notificationObservers.append(didEnterBackgroundObserver)

        // App will terminate
        let willTerminateObserver = notificationCenter.addObserver(
            forName: UIApplication.willTerminateNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppWillTerminate()
        }
        notificationObservers.append(willTerminateObserver)
        #elseif os(macOS)
        let notificationCenter = NotificationCenter.default

        // macOS: App will become active
        let didBecomeActiveObserver = notificationCenter.addObserver(
            forName: NSApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppDidBecomeActive()
        }
        notificationObservers.append(didBecomeActiveObserver)

        // macOS: App will resign active
        let willResignActiveObserver = notificationCenter.addObserver(
            forName: NSApplication.willResignActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppWillResignActive()
        }
        notificationObservers.append(willResignActiveObserver)

        // macOS: App will terminate
        let willTerminateObserver = notificationCenter.addObserver(
            forName: NSApplication.willTerminateNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppWillTerminate()
        }
        notificationObservers.append(willTerminateObserver)
        #endif
    }

    private func setupMemoryPressureObservers() {
        #if os(iOS)
        let notificationCenter = NotificationCenter.default

        // Memory warning
        let memoryWarningObserver = notificationCenter.addObserver(
            forName: UIApplication.didReceiveMemoryWarningNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleMemoryWarning()
        }
        notificationObservers.append(memoryWarningObserver)
        #elseif os(macOS)
        // macOS doesn't have the same memory warning API
        // Could monitor process memory manually if needed
        print("[\(Self.TAG)] Memory pressure monitoring not implemented for macOS")
        #endif
    }

    // MARK: - Lifecycle Event Handlers

    private func handleAppWillBecomeActive() {
        print("[\(Self.TAG)] App will become active")
        // Prepare for active use
    }

    private func handleAppDidBecomeActive() {
        print("[\(Self.TAG)] App did become active")
        isActive = true

        // End background task if running (iOS only)
        #if os(iOS)
        if backgroundTaskIdentifier != .invalid {
            UIApplication.shared.endBackgroundTask(backgroundTaskIdentifier)
            backgroundTaskIdentifier = .invalid
        }
        #endif
    }

    private func handleAppWillResignActive() {
        print("[\(Self.TAG)] App will resign active")
        isActive = false
    }

    private func handleAppDidEnterBackground() {
        print("[\(Self.TAG)] App did enter background")

        #if os(iOS)
        // Start background task to allow cleanup (iOS only)
        backgroundTaskIdentifier = UIApplication.shared.beginBackgroundTask(
            withName: "ExecutorchCleanup"
        ) { [weak self] in
            self?.handleBackgroundTaskExpiration()
        }
        #endif

        // Perform background cleanup
        Task {
            await performBackgroundCleanup()
        }
    }

    private func handleAppWillTerminate() {
        print("[\(Self.TAG)] App will terminate")
        Task {
            await disposeAllModels()
        }
    }

    private func handleBackgroundTaskExpiration() {
        print("[\(Self.TAG)] Background task expired")
        #if os(iOS)
        if backgroundTaskIdentifier != .invalid {
            UIApplication.shared.endBackgroundTask(backgroundTaskIdentifier)
            backgroundTaskIdentifier = .invalid
        }
        #endif
    }

    // MARK: - Memory Management

    private func handleMemoryWarning() {
        let now = Date()

        // Avoid too frequent memory warning handling
        if let lastWarning = lastMemoryWarning,
           now.timeIntervalSince(lastWarning) < memoryWarningCooldown {
            print("[\(Self.TAG)] Memory warning ignored (too frequent)")
            return
        }

        lastMemoryWarning = now
        print("[\(Self.TAG)] Memory warning received, performing cleanup")

        Task {
            await handleMemoryPressure()
        }
    }

    private func handleMemoryPressure() async {
        let memoryStats = getMemoryStats()
        print("[\(Self.TAG)] Memory pressure - used: \(memoryStats.usedBytes / 1024 / 1024)MB")

        // Dispose least recently used models
        let activeManagers = modelManagers.compactMap { $0.manager }
        for manager in activeManagers {
            // Let each manager handle its own memory pressure response
            await manager.handleMemoryPressure()
        }

        // Log memory usage after cleanup
        let newMemoryStats = getMemoryStats()
        print("[\(Self.TAG)] After cleanup - used: \(newMemoryStats.usedBytes / 1024 / 1024)MB")
    }

    private func performBackgroundCleanup() async {
        print("[\(Self.TAG)] Performing background cleanup")

        // Dispose non-essential models to free memory
        let activeManagers = modelManagers.compactMap { $0.manager }
        for manager in activeManagers {
            await manager.performBackgroundCleanup()
        }

        // End background task (iOS only)
        #if os(iOS)
        if backgroundTaskIdentifier != .invalid {
            UIApplication.shared.endBackgroundTask(backgroundTaskIdentifier)
            backgroundTaskIdentifier = .invalid
        }
        #endif
    }

    private func disposeAllModels() async {
        print("[\(Self.TAG)] Disposing all models")

        let activeManagers = modelManagers.compactMap { $0.manager }
        for manager in activeManagers {
            await manager.disposeAllModels()
        }

        modelManagers.removeAll()
    }

    // MARK: - Memory Monitoring Utilities

    private func getCurrentMemoryUsage() -> UInt64 {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size)/4

        let kerr: kern_return_t = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: 1) {
                task_info(mach_task_self_,
                         task_flavor_t(MACH_TASK_BASIC_INFO),
                         $0,
                         &count)
            }
        }

        return kerr == KERN_SUCCESS ? info.resident_size : 0
    }

    private func getMemoryPressureLevel() -> MemoryPressureLevel {
        let usedMemory = getCurrentMemoryUsage()
        let totalMemory = ProcessInfo.processInfo.physicalMemory
        let usageRatio = Double(usedMemory) / Double(totalMemory)

        switch usageRatio {
        case 0..<0.6:
            return .low
        case 0.6..<0.8:
            return .medium
        case 0.8..<0.95:
            return .high
        default:
            return .critical
        }
    }

    private func cleanup() {
        // Remove notification observers
        for observer in notificationObservers {
            NotificationCenter.default.removeObserver(observer)
        }
        notificationObservers.removeAll()

        // End background task if active (iOS only)
        #if os(iOS)
        if backgroundTaskIdentifier != .invalid {
            UIApplication.shared.endBackgroundTask(backgroundTaskIdentifier)
            backgroundTaskIdentifier = .invalid
        }
        #endif
    }
}

// MARK: - Extension for Model Manager Integration

extension ExecutorchModelManager {

    /**
     * Handle memory pressure by disposing least recently used models
     */
    func handleMemoryPressure() async {
        print("[\(Self.TAG)] Handling memory pressure")

        // In a real implementation, this would dispose LRU models
        // For now, we'll dispose half of the loaded models
        let loadedIds = try? getLoadedModelIds()
        let modelsToDispose = loadedIds?.prefix(loadedIds!.count / 2) ?? []

        for modelId in modelsToDispose {
            do {
                try await disposeModel(modelId: modelId)
                print("[\(Self.TAG)] Disposed model due to memory pressure: \(modelId)")
            } catch {
                print("[\(Self.TAG)] Failed to dispose model during memory pressure: \(modelId)")
            }
        }
    }

    /**
     * Perform background cleanup tasks
     */
    func performBackgroundCleanup() async {
        print("[\(Self.TAG)] Performing background cleanup")

        // In background, dispose non-essential models to free resources
        // This is a conservative approach to maintain app responsiveness
        await handleMemoryPressure()
    }
}

// MARK: - Supporting Data Types

/**
 * Memory usage statistics
 */
struct MemoryStats {
    let usedBytes: UInt64
    let availableBytes: UInt64
    let pressureLevel: MemoryPressureLevel
    let lastWarningTime: Date?

    var usageMB: Double {
        return Double(usedBytes) / (1024 * 1024)
    }

    var availableMB: Double {
        return Double(availableBytes) / (1024 * 1024)
    }

    var usagePercentage: Double {
        return availableBytes > 0 ? Double(usedBytes) / Double(availableBytes) * 100 : 0
    }
}

/**
 * Memory pressure levels
 */
enum MemoryPressureLevel {
    case low
    case medium
    case high
    case critical

    var description: String {
        switch self {
        case .low: return "Low"
        case .medium: return "Medium"
        case .high: return "High"
        case .critical: return "Critical"
        }
    }
}